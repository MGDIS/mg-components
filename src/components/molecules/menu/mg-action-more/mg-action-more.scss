@import '../../../../styles/a11y.scss';
