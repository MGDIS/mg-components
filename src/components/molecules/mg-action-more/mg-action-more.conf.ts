import { MgBadge } from '../../atoms/mg-badge/mg-badge';
import { MgButton } from '../../atoms/mg-button/mg-button';
import { MgIcon } from '../../atoms/mg-icon/mg-icon';
import { MgMenuItem } from '../menu/mg-menu-item/mg-menu-item';

/**
 * prop icon type
 */
export type MgActionMoreIconType = {
  icon: MgIcon['icon'];
};

/**
 * prop button type
 */
export type MgActionMoreButtonType = {
  isIcon: MgButton['isIcon'];
  variant: MgButton['variant'];
  label?: MgButton['label'];
};

/**
 * locals message type
 */
export type MgActionMoreMessageType = { label: string };

/**
 * prop item type
 */
export type MgActionMoreItemType = {
  label: string;
  mouseEventHandler: IMouseEventHandler;
  status?: MgMenuItem['status'];
  icon?: MgIcon['icon'];
  href?: MgMenuItem['href'];
  badge?: Pick<MgBadge, 'value' | 'label'>;
};

/**
 * MouseEvent callback interface
 */
interface IMouseEventHandler {
  (event: MouseEvent): void;
}
