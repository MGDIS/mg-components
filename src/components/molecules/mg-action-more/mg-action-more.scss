@import '../../../styles/a11y.scss';
@import '../../../styles/mixins/mg-icon-animation.scss';

:host {
  --mg-popover-padding-vertical: 1rem;
  --mg-popover-padding-horizontal: 0;
}

.mg-action-more__chevron {
  @include rotate('.mg-action-more__chevron', 200ms);
}
