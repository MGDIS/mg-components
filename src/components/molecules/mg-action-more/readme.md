## Anatomy

![](./mg-action-more/doc/img/mg-action-more-anatomy.png) 

## Specifications

![](./mg-action-more/doc/img/mg-action-more-popover-spacing.png) 

### Spacing
Popover padding should be 10px up/down, 0px left/right.

## Behavior

### Action

In addition of the standard popover's behavior, when an item of the menu is clicked the popover closes itself.

## Options

### Button's variant

![](./mg-action-more/doc/img/mg-action-more-options-variant.png) 

By default, button variant is "flat".
It is possible to set another variant (see [mg-button](http://core.pages.mgdis.fr/core-ui/mg-components/?path=/story/atoms-mg-button--mg-button)).

### Icon

![](./mg-action-more/doc/img/mg-action-more-options-icon.png)

By default, icon is *ellipsis*. 
It is possible to set another icon


### Label

![](./mg-action-more/doc/img/mg-action-more-options-label.png) 

By default, label is *Actions* and hidden. 
It is possible to set another label.

### Chevron

![](./mg-action-more/doc/img/mg-action-more-options-chevron.png) 

It is possible to display a chevron on the label right side, it make a 180 degree rotation on click.

<!-- Auto Generated Below -->


## Properties

| Property             | Attribute         | Description                                                        | Type                                                                                                                                             | Default                             |
| -------------------- | ----------------- | ------------------------------------------------------------------ | ------------------------------------------------------------------------------------------------------------------------------------------------ | ----------------------------------- |
| `button`             | --                | Define button properties Default: {variant: 'flat', isIcon: true}. | `{ isIcon: boolean; variant: "flat" \| "info" \| "success" \| "link" \| "primary" \| "secondary" \| "danger" \| "danger-alt"; label?: string; }` | `{ variant: 'flat', isIcon: true }` |
| `displayChevron`     | `display-chevron` | Define if chevron is display                                       | `boolean`                                                                                                                                        | `undefined`                         |
| `icon`               | --                | Define displaied icon Default: {icon: 'ellipsis'}                  | `{ icon: string; }`                                                                                                                              | `{ icon: 'ellipsis' }`              |
| `items` _(required)_ | --                | Define the menu-items elements                                     | `MgActionMoreItemType[]`                                                                                                                         | `undefined`                         |


## Dependencies

### Depends on

- [mg-icon](../../atoms/mg-icon)
- [mg-popover](../mg-popover)
- [mg-button](../../atoms/mg-button)
- [mg-menu](../menu/mg-menu)
- [mg-menu-item](../menu/mg-menu-item)
- [mg-badge](../../atoms/mg-badge)

### Graph
```mermaid
graph TD;
  mg-action-more --> mg-icon
  mg-action-more --> mg-popover
  mg-action-more --> mg-button
  mg-action-more --> mg-menu
  mg-action-more --> mg-menu-item
  mg-action-more --> mg-badge
  mg-popover --> mg-card
  mg-popover --> mg-button
  mg-popover --> mg-icon
  mg-button --> mg-icon
  mg-menu --> mg-item-more
  mg-item-more --> mg-menu-item
  mg-item-more --> mg-icon
  mg-item-more --> mg-menu
  mg-menu-item --> mg-badge
  mg-menu-item --> mg-icon
  mg-menu-item --> mg-popover
  style mg-action-more fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
