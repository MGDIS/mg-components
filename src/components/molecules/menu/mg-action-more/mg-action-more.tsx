import { Component, h, Element, Prop, Watch, State, EventEmitter, Event } from '@stencil/core';
import { sizes, Status } from '../mg-menu-item/mg-menu-item.conf';
import { initLocales } from '../../../../locales';
import { isMgActionMoreItem, ItemType, MessageType } from './mg-action-more.conf';
import { Direction } from '../mg-menu/mg-menu.conf';
import { MgIcon } from '../../../atoms/mg-icon/mg-icon';

@Component({
  tag: 'mg-action-more',
  styleUrl: 'mg-action-more.scss',
  shadow: true,
})
export class MgActionMore {
  /************
   * Internal *
   ************/
  private messages: MessageType;

  /**************
   * Decorators *
   **************/

  /**
   * Get component DOM element
   */
  @Element() element: HTMLMgActionMoreElement;

  /**
   * Define the menu-items elements
   */
  @Prop({ mutable: true }) items!: ItemType[];
  @Watch('items')
  validateItems(newValue: MgActionMore['items']) {
    if (typeof (newValue as MgActionMore['items']) === 'object' && newValue.find(item => !isMgActionMoreItem(item as ItemType) === undefined)) {
      throw new Error('<mg-action-more> prop "item" is required and all values must be the same type, ItemType.');
    } else {
      // update badge
      this.displayBadge = newValue.find(item => item.badge!) !== undefined;
    }
  }

  /**
   * Define if label is display
   */
  @Prop() displayLabel: boolean;

  /**
   * Define displaied icon
   * Default: 'ellipsis-vertical'
   */
  @Prop() icon: MgIcon['icon'] = 'ellipsis-vertical';

  /**
   * Emitted event when selected item change - item index number
   */
  @Event({ eventName: 'selected-change' }) selectedChange: EventEmitter<number>;

  /**********
   * States *
   **********/

  /**
   * Display if badge is display on more element
   */
  @State() displayBadge: boolean;

  /*************
   * Lifecycle *
   *************/

  /**
   * Validate props
   *
   * @returns {void}
   */
  componentWillLoad(): void {
    this.messages = initLocales(this.element).messages.plusMenu;
    this.validateItems(this.items);
  }

  /**
   * Render
   *
   * @returns {HTMLElement} HTML Element
   */
  render(): HTMLElement {
    return (
      <mg-popover>
        <mg-button variant="secondary">
          <mg-icon icon={this.icon}></mg-icon>
          <span class={{ 'sr-only': !this.displayLabel }}>{this.messages.moreLabel}</span>
          {this.displayBadge && <mg-badge label={this.messages.badgeLabel} value="!" hidden variant="primary"></mg-badge>}
        </mg-button>
        <div slot="content">
          <mg-menu direction={Direction.VERTICAL} label={this.messages.moreLabel}>
            {this.items.map((item, index) => (
              <mg-menu-item
                size={item.size || sizes[0]}
                hidden={item.hidden}
                status={item.status || Status.VISIBLE}
                onClick={() => {
                  if (item.proxyEvent) item.proxyEvent({ kind: 'click', payload: { bubbles: true } });
                  this.selectedChange.emit(index);
                }}
              >
                {item.icon && <mg-icon icon={item.icon} slot="image"></mg-icon>}
                <span slot="label">{item.label}</span>
                {item.badge?.label && <mg-badge label={item.badge.label} value={item.badge.value} slot="information" variant="primary"></mg-badge>}
              </mg-menu-item>
            ))}
          </mg-menu>
        </div>
      </mg-popover>
    );
  }
}
