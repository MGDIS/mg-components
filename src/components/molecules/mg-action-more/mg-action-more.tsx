import { Component, h, Element, Prop, Watch, State, Host } from '@stencil/core';
import { MgActionMoreItemType, MgActionMoreButtonType, MgActionMoreIconType, MgActionMoreMessageType } from './mg-action-more.conf';
import { initLocales } from '../../../locales';
import { Status } from '../menu/mg-menu-item/mg-menu-item.conf';
import { Direction } from '../menu/mg-menu/mg-menu.conf';
import { createID } from '../../../utils/components.utils';

/**
 * MgActionMore['items'] type guard
 *
 * @param {unknown} prop commponent items prop
 * @returns {boolean} return true if type is valid
 */
const isMgActionMoreItems = (prop: unknown): prop is MgActionMoreItemType[] => {
  const items = prop as MgActionMoreItemType[];
  return typeof items === 'object' && items.every(item => typeof item === 'object' && typeof item.label === 'string' && typeof item.mouseEventHandler === 'function');
};

/**
 * MgActionMore['button'] type guard
 *
 * @param {unknown} prop commponent button prop
 * @returns {boolean} return true if type is valid
 */
const isMgActionMoreButton = (prop: unknown): prop is MgActionMoreButtonType => {
  const button = prop as MgActionMoreButtonType;
  return typeof button === 'object' && typeof button.variant === 'string' && typeof button.isIcon === 'boolean';
};

/**
 * MgActionMore['icon'] type guard
 *
 * @param {unknown} prop commponent icon prop
 * @returns {boolean} return true if type is valid
 */
const isMgActionMoreIcon = (prop: unknown): prop is MgActionMoreIconType => {
  const icon = prop as MgActionMoreIconType;
  return typeof icon === 'object' && typeof icon.icon === 'string';
};

@Component({
  tag: 'mg-action-more',
  styleUrl: 'mg-action-more.scss',
  shadow: true,
})
export class MgActionMore {
  /************
   * Internal *
   ************/
  private readonly name = 'mg-action-more';
  private readonly mgPopoverIdentifier = createID(this.name);
  private messages: MgActionMoreMessageType;

  /**************
   * Decorators *
   **************/

  /**
   * Get component DOM element
   */
  @Element() element: HTMLMgActionMoreElement;

  /**
   * Define displaied icon
   * Default: {icon: 'ellipsis'}
   */
  @Prop() icon: MgActionMoreIconType = { icon: 'ellipsis' };
  @Watch('icon')
  validateIcon(newValue: MgActionMore['icon']): void {
    if (newValue && !isMgActionMoreIcon(newValue)) {
      throw new Error(`<${this.name}> prop icon must match MgActionMoreIconType.`);
    }
  }

  /**
   * Define button properties
   * Default: {variant: 'flat', isIcon: true}.
   */
  @Prop() button: MgActionMoreButtonType = { variant: 'flat', isIcon: true };
  @Watch('button')
  validateButton(newValue: MgActionMore['button']): void {
    if (newValue && !isMgActionMoreButton(newValue)) {
      throw new Error(`<${this.name}> prop button must match MgActionMoreButtonType.`);
    }
  }

  /**
   * Define the menu-items elements
   */
  @Prop() items!: MgActionMoreItemType[];
  @Watch('items')
  validateItems(newValue: MgActionMore['items']): void {
    if (!isMgActionMoreItems(newValue)) {
      throw new Error(`<${this.name}> prop "items" is required and all values must be the same type, MgActionMoreItemType.`);
    }
  }

  /**
   * Define if chevron is display
   */
  @Prop() displayChevron: boolean;
  @Watch('displayChevron')
  validateDisplayChevron(newValue: MgActionMore['displayChevron']): void {
    if (newValue && this.button.isIcon === true) {
      throw new Error(`<${this.name}> prop "displayChevron" can't be used with a 'button" prop "isIcon" attribute.`);
    }
  }
  /**
   * Menu expanded state
   * Default: false.
   */
  @State() expanded = false;

  /************
   * Merthods *
   ***********/

  /**
   * Toogle expanded props
   *
   * @returns {void}
   */
  private toggleExpanded = (): void => {
    this.expanded = !this.expanded;
  };

  /**
   * Button click handler
   *
   * @returns {void}
   */
  private handleButton = (): void => {
    this.toggleExpanded();
  };

  /**
   * Item click handler
   *
   * @param {MouseEvent} event click event
   * @param {MgActionMoreItemType['mouseEventHandler']} customHandler item['mouseEventHandler']
   * @returns {void}
   */
  private handleItemClick = (event: MouseEvent, customHandler: MgActionMoreItemType['mouseEventHandler']) => {
    this.expanded = false;
    customHandler(event);
  };

  /*************
   * Lifecycle *
   *************/

  /**
   * Validate props
   *
   * @returns {void}
   */
  componentWillLoad(): void {
    this.messages = initLocales(this.element).messages.actionMore as MgActionMoreMessageType;
    this.validateItems(this.items);
    this.validateButton(this.button);
    this.validateIcon(this.icon);
    this.validateDisplayChevron(this.displayChevron);
  }

  /**
   * Render
   *
   * @returns {HTMLElement} HTML Element
   */
  render(): HTMLElement {
    const buttonLabel = this.button.label || this.messages.label;
    const buttonContent = [buttonLabel];

    if (!this.button.isIcon && this.displayChevron) {
      buttonContent.push(
        <span
          class={{
            [`${this.name}__chevron`]: true,
            [`${this.name}__chevron--rotate`]: this.expanded === true,
            'mg-a11y-animation': true,
          }}
        >
          <mg-icon icon="chevron-down" size="small"></mg-icon>
        </span>,
      );
    }

    return (
      <Host data-mg-popover-guard={this.mgPopoverIdentifier}>
        <span class="mg-action-more">
          <mg-popover identifier={this.mgPopoverIdentifier} display={this.expanded}>
            <mg-button variant={this.button.variant} isIcon={this.button.isIcon} type="button" label={buttonLabel} onClick={this.handleButton}>
              <mg-icon {...this.icon}></mg-icon>
              {!this.button.isIcon && buttonContent}
            </mg-button>
            <div slot="content">
              <mg-menu direction={Direction.VERTICAL} label={this.messages.label}>
                {this.items.map(item => (
                  // eslint-disable-next-line react/jsx-no-bind
                  <mg-menu-item key={item.label} status={item.status || Status.VISIBLE} onClick={e => this.handleItemClick(e, item.mouseEventHandler)} href={item.href}>
                    {item.icon && <mg-icon icon={item.icon} slot="image"></mg-icon>}
                    <span slot="label">{item.label}</span>
                    {item.badge?.label && <mg-badge label={item.badge.label} value={item.badge.value} slot="information" variant="text-color"></mg-badge>}
                  </mg-menu-item>
                ))}
              </mg-menu>
            </div>
          </mg-popover>
        </span>
      </Host>
    );
  }
}
