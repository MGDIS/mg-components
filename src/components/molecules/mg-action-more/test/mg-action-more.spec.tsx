import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';
import { Status } from '../../menu/mg-menu-item/mg-menu-item.conf';
import { MgActionMore } from '../mg-action-more';

const getPage = args =>
  newSpecPage({
    components: [MgActionMore],
    template: () => <mg-action-more {...args}></mg-action-more>,
  });

const mouseEventHandler = jest.fn();

const items = [
  {
    label: 'batman',
    mouseEventHandler,
  },
  {
    label: 'robin',
    mouseEventHandler,
    status: Status.HIDDEN,
  },
  {
    label: 'joker',
    mouseEventHandler,
    badge: {
      value: 2,
      label: 'badge',
    },
  },
  {
    label: 'bane',
    mouseEventHandler,
    icon: 'user',
  },
];

describe('mg-action-more', () => {
  describe.each([
    {
      items,
    },
    {
      items,
      button: {
        label: 'dc-comics',
        variant: 'primary',
        isIcon: false,
      },
    },
    {
      items,
      displayChevron: true,
      button: {
        variant: 'flat',
        isIcon: false,
      },
    },
  ])('render', args => {
    test(`Should render with args ${args}`, async () => {
      const { root } = await getPage(args);
      expect(root).toMatchSnapshot();
    });

    test('Should toggle menu on click', async () => {
      const page = await getPage(args);
      expect(page.root).toMatchSnapshot();

      const mgMoreAction = page.doc.querySelector('mg-action-more');
      const mgButton = mgMoreAction.shadowRoot.querySelector('mg-button');
      mgButton.dispatchEvent(new MouseEvent('click', { bubbles: true }));

      await page.waitForChanges();

      expect(page.root).toMatchSnapshot();
    });
  });

  describe('errors', () => {
    test.each([
      { args: {}, error: `<mg-action-more> prop "item" is required and all values must be the same type, MgActionMoreItemType.` },
      { args: { items: ['batman'] }, error: `<mg-action-more> prop "item" is required and all values must be the same type, MgActionMoreItemType.` },
      { args: { items: [{ label: 'batman' }] }, error: `<mg-action-more> prop "item" is required and all values must be the same type, MgActionMoreItemType.` },
      {
        args: { items: [{ label: 'batman', mouseEventHandler: 'batman' }] },
        error: `<mg-action-more> prop "item" is required and all values must be the same type, MgActionMoreItemType.`,
      },
      { args: { items: [{ mouseEventHandler: 'batman' }] }, error: `<mg-action-more> prop "item" is required and all values must be the same type, MgActionMoreItemType.` },
      { args: { items: [{ label: 'batman', mouseEventHandler }], button: {} }, error: `<mg-action-more> prop button must match MgActionMoreButtonType.` },
      { args: { items: [{ label: 'batman', mouseEventHandler }], button: { variant: 'primary' } }, error: `<mg-action-more> prop button must match MgActionMoreButtonType.` },
      { args: { items: [{ label: 'batman', mouseEventHandler }], button: { isIcon: true } }, error: `<mg-action-more> prop button must match MgActionMoreButtonType.` },
      { args: { items: [{ label: 'batman', mouseEventHandler }], icon: 'user' }, error: `<mg-action-more> prop icon must match MgActionMoreIconType.` },
      {
        args: { items: [{ label: 'batman', mouseEventHandler }], displayChevron: true },
        error: `<mg-action-more> prop "displayChevron" can't be used with a 'button" prop "isIcon" attribute.`,
      },
      {
        args: { items: [{ label: 'batman', mouseEventHandler }], button: { isIcon: true, variant: 'flat' }, displayChevron: true },
        error: `<mg-action-more> prop "displayChevron" can't be used with a 'button" prop "isIcon" attribute.`,
      },
    ])('with args %s', async ({ args, error }) => {
      expect.assertions(1);
      try {
        await getPage(args);
      } catch (err) {
        expect(err.message).toMatch(error);
      }
    });
  });

  describe('navigation', () => {
    test(`Should run item mouseEventHandler by clicking on menu-item`, async () => {
      const page = await getPage({ items });
      expect(page.root).toMatchSnapshot();

      const mgMoreAction = page.doc.querySelector('mg-action-more');
      const mgButton = mgMoreAction.shadowRoot.querySelector('mg-button');
      mgButton.dispatchEvent(new MouseEvent('click', { bubbles: true }));

      await page.waitForChanges();

      expect(page.root).toMatchSnapshot();

      const mgMenuItem = mgMoreAction.shadowRoot.querySelector('mg-menu-item');
      mgMenuItem.dispatchEvent(new MouseEvent('click', { bubbles: true }));

      await page.waitForChanges();

      expect(mouseEventHandler).toHaveBeenCalled();
      expect(page.root).toMatchSnapshot();
    });
  });
});
