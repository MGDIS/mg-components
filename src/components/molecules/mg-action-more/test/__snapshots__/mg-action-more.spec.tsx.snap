// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mg-action-more navigation Should run item mouseEventHandler by clicking on menu-item 1`] = `
<mg-action-more>
  <mock:shadow-root>
    <span class="mg-action-more">
      <mg-popover>
        <mg-button isicon="" label="Actions" type="button" variant="flat">
          <mg-icon icon="ellipsis"></mg-icon>
        </mg-button>
        <div slot="content">
          <mg-menu direction="vertical" label="Actions">
            <mg-menu-item status="visible">
              <span slot="label">
                batman
              </span>
            </mg-menu-item>
            <mg-menu-item status="hidden">
              <span slot="label">
                robin
              </span>
            </mg-menu-item>
            <mg-menu-item status="visible">
              <span slot="label">
                joker
              </span>
              <mg-badge label="badge" slot="information" value="2" variant="text-color"></mg-badge>
            </mg-menu-item>
            <mg-menu-item status="visible">
              <mg-icon icon="user" slot="image"></mg-icon>
              <span slot="label">
                bane
              </span>
            </mg-menu-item>
          </mg-menu>
        </div>
      </mg-popover>
    </span>
  </mock:shadow-root>
</mg-action-more>
`;

exports[`mg-action-more navigation Should run item mouseEventHandler by clicking on menu-item 2`] = `
<mg-action-more>
  <mock:shadow-root>
    <span class="mg-action-more">
      <mg-popover>
        <mg-button isicon="" label="Actions" type="button" variant="flat">
          <mg-icon icon="ellipsis"></mg-icon>
        </mg-button>
        <div slot="content">
          <mg-menu direction="vertical" label="Actions">
            <mg-menu-item status="visible">
              <span slot="label">
                batman
              </span>
            </mg-menu-item>
            <mg-menu-item status="hidden">
              <span slot="label">
                robin
              </span>
            </mg-menu-item>
            <mg-menu-item status="visible">
              <span slot="label">
                joker
              </span>
              <mg-badge label="badge" slot="information" value="2" variant="text-color"></mg-badge>
            </mg-menu-item>
            <mg-menu-item status="visible">
              <mg-icon icon="user" slot="image"></mg-icon>
              <span slot="label">
                bane
              </span>
            </mg-menu-item>
          </mg-menu>
        </div>
      </mg-popover>
    </span>
  </mock:shadow-root>
</mg-action-more>
`;

exports[`mg-action-more navigation Should run item mouseEventHandler by clicking on menu-item 3`] = `
<mg-action-more>
  <mock:shadow-root>
    <span class="mg-action-more">
      <mg-popover>
        <mg-button isicon="" label="Actions" type="button" variant="flat">
          <mg-icon icon="ellipsis"></mg-icon>
        </mg-button>
        <div slot="content">
          <mg-menu direction="vertical" label="Actions">
            <mg-menu-item status="visible">
              <span slot="label">
                batman
              </span>
            </mg-menu-item>
            <mg-menu-item status="hidden">
              <span slot="label">
                robin
              </span>
            </mg-menu-item>
            <mg-menu-item status="visible">
              <span slot="label">
                joker
              </span>
              <mg-badge label="badge" slot="information" value="2" variant="text-color"></mg-badge>
            </mg-menu-item>
            <mg-menu-item status="visible">
              <mg-icon icon="user" slot="image"></mg-icon>
              <span slot="label">
                bane
              </span>
            </mg-menu-item>
          </mg-menu>
        </div>
      </mg-popover>
    </span>
  </mock:shadow-root>
</mg-action-more>
`;

exports[`mg-action-more render Should render with args [object Object] 1`] = `
<mg-action-more>
  <mock:shadow-root>
    <span class="mg-action-more">
      <mg-popover>
        <mg-button isicon="" label="Actions" type="button" variant="flat">
          <mg-icon icon="ellipsis"></mg-icon>
        </mg-button>
        <div slot="content">
          <mg-menu direction="vertical" label="Actions">
            <mg-menu-item status="visible">
              <span slot="label">
                batman
              </span>
            </mg-menu-item>
            <mg-menu-item status="hidden">
              <span slot="label">
                robin
              </span>
            </mg-menu-item>
            <mg-menu-item status="visible">
              <span slot="label">
                joker
              </span>
              <mg-badge label="badge" slot="information" value="2" variant="text-color"></mg-badge>
            </mg-menu-item>
            <mg-menu-item status="visible">
              <mg-icon icon="user" slot="image"></mg-icon>
              <span slot="label">
                bane
              </span>
            </mg-menu-item>
          </mg-menu>
        </div>
      </mg-popover>
    </span>
  </mock:shadow-root>
</mg-action-more>
`;

exports[`mg-action-more render Should render with args [object Object] 2`] = `
<mg-action-more>
  <mock:shadow-root>
    <span class="mg-action-more">
      <mg-popover>
        <mg-button label="dc-comics" type="button" variant="primary">
          <mg-icon icon="ellipsis"></mg-icon>
          dc-comics
        </mg-button>
        <div slot="content">
          <mg-menu direction="vertical" label="Actions">
            <mg-menu-item status="visible">
              <span slot="label">
                batman
              </span>
            </mg-menu-item>
            <mg-menu-item status="hidden">
              <span slot="label">
                robin
              </span>
            </mg-menu-item>
            <mg-menu-item status="visible">
              <span slot="label">
                joker
              </span>
              <mg-badge label="badge" slot="information" value="2" variant="text-color"></mg-badge>
            </mg-menu-item>
            <mg-menu-item status="visible">
              <mg-icon icon="user" slot="image"></mg-icon>
              <span slot="label">
                bane
              </span>
            </mg-menu-item>
          </mg-menu>
        </div>
      </mg-popover>
    </span>
  </mock:shadow-root>
</mg-action-more>
`;

exports[`mg-action-more render Should render with args [object Object] 3`] = `
<mg-action-more>
  <mock:shadow-root>
    <span class="mg-action-more">
      <mg-popover>
        <mg-button label="Actions" type="button" variant="flat">
          <mg-icon icon="ellipsis"></mg-icon>
          Actions
          <span class="mg-a11y-animation mg-action-more__chevron">
            <mg-icon icon="chevron-down" size="small"></mg-icon>
          </span>
        </mg-button>
        <div slot="content">
          <mg-menu direction="vertical" label="Actions">
            <mg-menu-item status="visible">
              <span slot="label">
                batman
              </span>
            </mg-menu-item>
            <mg-menu-item status="hidden">
              <span slot="label">
                robin
              </span>
            </mg-menu-item>
            <mg-menu-item status="visible">
              <span slot="label">
                joker
              </span>
              <mg-badge label="badge" slot="information" value="2" variant="text-color"></mg-badge>
            </mg-menu-item>
            <mg-menu-item status="visible">
              <mg-icon icon="user" slot="image"></mg-icon>
              <span slot="label">
                bane
              </span>
            </mg-menu-item>
          </mg-menu>
        </div>
      </mg-popover>
    </span>
  </mock:shadow-root>
</mg-action-more>
`;

exports[`mg-action-more render Should toggle menu on click 1`] = `
<mg-action-more>
  <mock:shadow-root>
    <span class="mg-action-more">
      <mg-popover>
        <mg-button isicon="" label="Actions" type="button" variant="flat">
          <mg-icon icon="ellipsis"></mg-icon>
        </mg-button>
        <div slot="content">
          <mg-menu direction="vertical" label="Actions">
            <mg-menu-item status="visible">
              <span slot="label">
                batman
              </span>
            </mg-menu-item>
            <mg-menu-item status="hidden">
              <span slot="label">
                robin
              </span>
            </mg-menu-item>
            <mg-menu-item status="visible">
              <span slot="label">
                joker
              </span>
              <mg-badge label="badge" slot="information" value="2" variant="text-color"></mg-badge>
            </mg-menu-item>
            <mg-menu-item status="visible">
              <mg-icon icon="user" slot="image"></mg-icon>
              <span slot="label">
                bane
              </span>
            </mg-menu-item>
          </mg-menu>
        </div>
      </mg-popover>
    </span>
  </mock:shadow-root>
</mg-action-more>
`;

exports[`mg-action-more render Should toggle menu on click 2`] = `
<mg-action-more>
  <mock:shadow-root>
    <span class="mg-action-more">
      <mg-popover>
        <mg-button isicon="" label="Actions" type="button" variant="flat">
          <mg-icon icon="ellipsis"></mg-icon>
        </mg-button>
        <div slot="content">
          <mg-menu direction="vertical" label="Actions">
            <mg-menu-item status="visible">
              <span slot="label">
                batman
              </span>
            </mg-menu-item>
            <mg-menu-item status="hidden">
              <span slot="label">
                robin
              </span>
            </mg-menu-item>
            <mg-menu-item status="visible">
              <span slot="label">
                joker
              </span>
              <mg-badge label="badge" slot="information" value="2" variant="text-color"></mg-badge>
            </mg-menu-item>
            <mg-menu-item status="visible">
              <mg-icon icon="user" slot="image"></mg-icon>
              <span slot="label">
                bane
              </span>
            </mg-menu-item>
          </mg-menu>
        </div>
      </mg-popover>
    </span>
  </mock:shadow-root>
</mg-action-more>
`;

exports[`mg-action-more render Should toggle menu on click 3`] = `
<mg-action-more>
  <mock:shadow-root>
    <span class="mg-action-more">
      <mg-popover>
        <mg-button label="dc-comics" type="button" variant="primary">
          <mg-icon icon="ellipsis"></mg-icon>
          dc-comics
        </mg-button>
        <div slot="content">
          <mg-menu direction="vertical" label="Actions">
            <mg-menu-item status="visible">
              <span slot="label">
                batman
              </span>
            </mg-menu-item>
            <mg-menu-item status="hidden">
              <span slot="label">
                robin
              </span>
            </mg-menu-item>
            <mg-menu-item status="visible">
              <span slot="label">
                joker
              </span>
              <mg-badge label="badge" slot="information" value="2" variant="text-color"></mg-badge>
            </mg-menu-item>
            <mg-menu-item status="visible">
              <mg-icon icon="user" slot="image"></mg-icon>
              <span slot="label">
                bane
              </span>
            </mg-menu-item>
          </mg-menu>
        </div>
      </mg-popover>
    </span>
  </mock:shadow-root>
</mg-action-more>
`;

exports[`mg-action-more render Should toggle menu on click 4`] = `
<mg-action-more>
  <mock:shadow-root>
    <span class="mg-action-more">
      <mg-popover>
        <mg-button label="dc-comics" type="button" variant="primary">
          <mg-icon icon="ellipsis"></mg-icon>
          dc-comics
        </mg-button>
        <div slot="content">
          <mg-menu direction="vertical" label="Actions">
            <mg-menu-item status="visible">
              <span slot="label">
                batman
              </span>
            </mg-menu-item>
            <mg-menu-item status="hidden">
              <span slot="label">
                robin
              </span>
            </mg-menu-item>
            <mg-menu-item status="visible">
              <span slot="label">
                joker
              </span>
              <mg-badge label="badge" slot="information" value="2" variant="text-color"></mg-badge>
            </mg-menu-item>
            <mg-menu-item status="visible">
              <mg-icon icon="user" slot="image"></mg-icon>
              <span slot="label">
                bane
              </span>
            </mg-menu-item>
          </mg-menu>
        </div>
      </mg-popover>
    </span>
  </mock:shadow-root>
</mg-action-more>
`;

exports[`mg-action-more render Should toggle menu on click 5`] = `
<mg-action-more>
  <mock:shadow-root>
    <span class="mg-action-more">
      <mg-popover>
        <mg-button label="Actions" type="button" variant="flat">
          <mg-icon icon="ellipsis"></mg-icon>
          Actions
          <span class="mg-a11y-animation mg-action-more__chevron">
            <mg-icon icon="chevron-down" size="small"></mg-icon>
          </span>
        </mg-button>
        <div slot="content">
          <mg-menu direction="vertical" label="Actions">
            <mg-menu-item status="visible">
              <span slot="label">
                batman
              </span>
            </mg-menu-item>
            <mg-menu-item status="hidden">
              <span slot="label">
                robin
              </span>
            </mg-menu-item>
            <mg-menu-item status="visible">
              <span slot="label">
                joker
              </span>
              <mg-badge label="badge" slot="information" value="2" variant="text-color"></mg-badge>
            </mg-menu-item>
            <mg-menu-item status="visible">
              <mg-icon icon="user" slot="image"></mg-icon>
              <span slot="label">
                bane
              </span>
            </mg-menu-item>
          </mg-menu>
        </div>
      </mg-popover>
    </span>
  </mock:shadow-root>
</mg-action-more>
`;

exports[`mg-action-more render Should toggle menu on click 6`] = `
<mg-action-more>
  <mock:shadow-root>
    <span class="mg-action-more">
      <mg-popover>
        <mg-button label="Actions" type="button" variant="flat">
          <mg-icon icon="ellipsis"></mg-icon>
          Actions
          <span class="mg-a11y-animation mg-action-more__chevron mg-action-more__chevron--rotate">
            <mg-icon icon="chevron-down" size="small"></mg-icon>
          </span>
        </mg-button>
        <div slot="content">
          <mg-menu direction="vertical" label="Actions">
            <mg-menu-item status="visible">
              <span slot="label">
                batman
              </span>
            </mg-menu-item>
            <mg-menu-item status="hidden">
              <span slot="label">
                robin
              </span>
            </mg-menu-item>
            <mg-menu-item status="visible">
              <span slot="label">
                joker
              </span>
              <mg-badge label="badge" slot="information" value="2" variant="text-color"></mg-badge>
            </mg-menu-item>
            <mg-menu-item status="visible">
              <mg-icon icon="user" slot="image"></mg-icon>
              <span slot="label">
                bane
              </span>
            </mg-menu-item>
          </mg-menu>
        </div>
      </mg-popover>
    </span>
  </mock:shadow-root>
</mg-action-more>
`;
